crates/kitsune_p2p/types/src/dht_arc.rs

```diff
@@ -172,147 +389,259 @@ impl RangeBounds<u32> for ArcRange {
-            assert!(DhtArc::new(mid, hl).contains(end));
-            assert!(!DhtArc::new(mid, hl).contains(out_r));
-            assert!(!DhtArc::new(mid, hl + 1).contains(opp));
-        };
+impl DhtArcBucket {
+    /// Select only the arcs that fit into the bucket.
+    pub fn new<I: IntoIterator<Item = DhtArc>>(filter: DhtArc, arcs: I) -> Self {
+        let arcs = arcs
+            .into_iter()
+            .filter(|a| filter.contains(a.center_loc))
+            .collect();
+        Self { filter, arcs }
+    }
 
-        // Checks that everything is contained because this is a full range
-        let check_bounds_full = |mid, hl, start, end| {
-            let out_l = (Wrapping(start) - Wrapping(1u32)).0;
-            let out_r = (Wrapping(end) + Wrapping(1u32)).0;
-            let opp = (Wrapping(mid) + Wrapping(half)).0;
-
-            assert!(DhtArc::new(mid, hl).contains(out_l));
-            assert!(DhtArc::new(mid, hl).contains(start));
-            assert!(DhtArc::new(mid, hl).contains(mid));
-            assert!(DhtArc::new(mid, hl).contains(end));
-            assert!(DhtArc::new(mid, hl).contains(out_r));
-            assert!(DhtArc::new(mid, hl + 1).contains(opp));
-        };
+    /// Same as new but doesn't check if arcs fit into the bucket.
+    pub fn new_unchecked(bucket: DhtArc, arcs: Vec<DhtArc>) -> Self {
+        Self {
+            filter: bucket,
+            arcs,
+        }
+    }
 
-        assert!(DhtArc::new(0, 0).range().is_empty());
-        assert_eq!(DhtArc::new(0, 1).range().into_inc(), 0..=0);
-        assert_eq!(DhtArc::new(1, 2).range().into_inc(), 0..=2);
-        assert_eq!(
-            DhtArc::new(quarter, quarter + 1).range().into_inc(),
-            0..=(half as usize)
-        );
-        check_bounds(quarter, quarter + 1, 0, half);
-
-        assert_eq!(
-            DhtArc::new(half, quarter + 1).range().into_inc(),
-            (quarter as usize)..=((quarter * 3) as usize)
-        );
-        check_bounds(half, quarter + 1, quarter, quarter * 3);
-
-        assert_eq!(
-            DhtArc::new(half, MAX_HALF_LENGTH).range().into_inc(),
-            0..=(u32::MAX as usize)
-        );
-        check_bounds_full(half, MAX_HALF_LENGTH, 0, u32::MAX);
-
-        assert_eq!(
-            DhtArc::new(half, MAX_HALF_LENGTH - 2).range().into_inc(),
-            2..=((u32::MAX - 1) as usize)
-        );
-        check_bounds(half, MAX_HALF_LENGTH - 2, 2, u32::MAX - 1);
-
-        assert_eq!(
-            DhtArc::new(0, 2).range(),
-            ArcRange {
-                start: Included(u32::MAX),
-                end: Included(1)
+    /// Get the density of this bucket.
+    pub fn density(&self) -> PeerDensity {
+        self.arcs
+            .iter()
+            .fold(PeerDensity::new(self.filter, 0.0, 0), |mut d, arc| {
+                d.average_coverage =
+                    (d.average_coverage * d.count as f64 + arc.coverage()) / (d.count as f64 + 1.0);
+                d.count += 1;
```

Comment:
tbh i think this is totally sufficient as f64 division errors are minuscule, but outside code reviewers will look at this and say we should add all the coverages together first and divide once by the count at the end...

Reply:
Interesting. I can make that change.